### ISSUE TITLE
`macos_command_line_application`: Info.plist is only embedded in the first `--macos_cpus` slice of a universal binary (regression in 4.4.0)

### ISSUE THREAD

**Issue by mlw** [bazelbuild/rules_apple#3041]
## Summary

Since 4.4.0, a universal `macos_command_line_application` only gets its
`__TEXT,__info_plist` section in one of its slices: the architecture listed first in
`--macos_cpus`. The other slice is linked without the `-sectcreate` linkopt, so it
has no embedded Info.plist. Single-architecture builds of either architecture are
fine.

`codesign` cannot repair this, because the plist is embedded by the linker.
Re-signing produces a valid signature over a slice that still has no plist, and
Apple's notary service then rejects the binary with `The signature of the binary is
invalid.`. `codesign --verify` without `--arch` does not catch it either, since it
only checks the slice matching the host.

Last good version is 4.3.3. Still present on `main` (bf7e682b).

## Reproducer

```python
# MODULE.bazel
module(name = "plistrepro", version = "0.0.0")

# apple_support must be listed above rules_cc so the Apple CC toolchain is
# registered first (see apple_support README).
bazel_dep(name = "apple_support", version = "2.8.0")
bazel_dep(name = "rules_cc", version = "0.2.22")
bazel_dep(name = "rules_apple", version = "4.5.3")
```

```python
# BUILD
load("@rules_apple//apple:macos.bzl", "macos_command_line_application")
load("@rules_cc//cc:cc_library.bzl", "cc_library")

cc_library(
    name = "lib",
    srcs = ["main.c"],
)

macos_command_line_application(
    name = "tool",
    bundle_id = "com.example.tool",
    infoplists = ["Info.plist"],
    minimum_os_version = "14.0",
    deps = [":lib"],
)
```

`main.c` is `int main(void) { return 0; }`, and `Info.plist` contains only
`CFBundleIdentifier` set to `com.example.tool`.

## Observed

```console
$ bazel build -c opt --macos_cpus=arm64,x86_64 //:tool

$ lipo -archs bazel-bin/tool
x86_64 arm64

$ otool -arch x86_64 -s __TEXT __info_plist bazel-bin/tool
bazel-bin/tool:
                                          # section absent entirely

$ otool -arch arm64 -s __TEXT __info_plist bazel-bin/tool
bazel-bin/tool:
Contents of (__TEXT,__info_plist) section
0000000100001438	696c7062 30307473 030201dc 07060504
...
```

Same thing as `codesign` sees it:

```console
$ codesign -d --arch x86_64 -vvv bazel-bin/tool 2>&1 | grep -i info.plist
Info.plist=not bound

$ codesign -d --arch arm64 -vvv bazel-bin/tool 2>&1 | grep -i info.plist
Info.plist entries=12
```

Expected: both slices carry `__TEXT,__info_plist`, as they do on 4.3.3.

## It is the first-listed CPU that wins, not the host

Reversing `--macos_cpus` moves the defect to the other slice. Both runs below are on
the same arm64 host:

| Build | arm64 slice | x86_64 slice |
|---|---|---|
| `--macos_cpus=arm64,x86_64` | present | **missing** |
| `--macos_cpus=x86_64,arm64` | **missing** | present |

So no ordering of `--macos_cpus` produces a correct universal binary. Building either
architecture on its own (`--macos_cpus=x86_64` or `--macos_cpus=arm64`) is fine, so
this is specific to the split build.

## Versions affected

Same reproducer, `--macos_cpus=arm64,x86_64`, arm64 host:

| rules_apple | arm64 slice | x86_64 slice |
|---|---|---|
| 4.3.3 | present | present |
| 4.4.0 | present | **missing** |
| 4.5.3 | present | **missing** |
| 5.0.0-rc2 | present | **missing** |
| main (bf7e682b) | present | **missing** |

4.5.0, 4.5.1 and 4.5.2 are affected too. 4.4.0 is the first bad version and the
immediate successor to 4.3.3.

## Root cause

The merged-plist target is analyzed in one configuration instead of once per split.

`macos_command_line_application` is a macro that creates a `macos_binary_infoplist`
target (`<name>.merged_infoplist`) and adds it to the binary's `deps`. That target
propagates the `-sectcreate __TEXT __info_plist` linkopt into the link
(`apple/macos.bzl`, `linking_support.sectcreate_objc_provider`).

`deps` is split by `--macos_cpus`, so on 4.3.3 the plist target is configured once per
split. On 4.5.3 it is configured only once in total:

```console
$ bazel cquery -c opt --macos_cpus=arm64,x86_64 'deps(//:tool)' --output=label_kind \
    | grep merged_infoplist

# 4.3.3
macos_binary_infoplist rule //:tool.merged_infoplist (1d31844)
macos_binary_infoplist rule //:tool.merged_infoplist (f0a559d)

# 4.5.3
macos_binary_infoplist rule //:tool.merged_infoplist (f0a559d)
```

The linkopt therefore reaches only one of the two link actions:

```console
$ bazel aquery -c opt --macos_cpus=arm64,x86_64 'mnemonic("ObjcLink|CppLink", //:tool)'
```

| link action output tree | 4.3.3 | 4.5.3 |
|---|---|---|
| `darwin_arm64-.../bin/tool_bin` | `-Wl,-sectcreate,__TEXT,__info_plist,...` | `-Wl,-sectcreate,__TEXT,__info_plist,...` |
| `darwin_x86_64-.../bin/tool_bin` | `-Wl,-sectcreate,__TEXT,__info_plist,...` | *(absent)* |

On 4.5.3 the x86_64 link action also has no merged `Info.plist` among its inputs.

The change that introduced this appears to be #2846 ("Migrate uses of
`ctx.fragments.apple.{single/multi}_arch_platform(...)` to platform constraints"),
the only commit touching `apple/internal/macos_binary_support.bzl` between 4.3.3 and
4.4.0. It gave `macos_binary_infoplist` its own incoming rule transition:

```diff
 macos_binary_infoplist = rule(
     implementation = _macos_binary_infoplist_impl,
+    cfg = transition_support.apple_rule_transition,
     attrs = dicts.add(
+        apple_support.platform_constraint_attrs(),
         rule_attrs.common_tool_attrs(),
+        rule_attrs.custom_transition_allowlist_attr(),
```

`apple_rule_transition` is `_apple_rule_base_transition`, which keeps only the first
architecture (`apple/internal/transition_support.bzl`):

```python
def _apple_rule_base_transition_impl(settings, attr):
    """Rule transition for Apple rules using Bazel CPUs and a valid Apple split transition."""
    minimum_os_version = attr.minimum_os_version
    platform_type = attr.platform_type
    return _command_line_options(
        environment_arch = _environment_archs(platform_type, minimum_os_version, settings)[0],
        ...
```

That `[0]` matches the observed behavior. `macos_binary_infoplist` is reached
*through* the binary rule's `--macos_cpus` split, but its own 1:1 incoming transition
collapses it to a single configuration chosen by the first CPU, and only the split
matching that configuration ends up with the linkopt.

The transition looks correct on the bundle rules that use it, where the rule is the
top-level target that picks one configuration and splits internally. The problem is
applying it to a rule that is itself a dependency inside a split.

## Also affects `launchdplists` and `apple_universal_binary`

`macos_command_line_launchdplist` got the same `cfg = transition_support.apple_rule_transition`
in the same commit and behaves identically: with both `infoplists` and `launchdplists`
set on 4.5.3, the arm64 slice has `__TEXT,__info_plist` and `__TEXT,__launchd_plist`
and the x86_64 slice has neither.

Wrapping the tool in `apple_universal_binary`, which the docs describe as "primarily
useful to force macOS tools to be built as universal binaries using
`forced_cpus = ["x86_64", "arm64"]`", hits the same defect, with the first entry of
`forced_cpus` keeping its plist:

```python
apple_universal_binary(
    name = "tool_universal",
    binary = ":tool",
    forced_cpus = ["x86_64", "arm64"],
    minimum_os_version = "14.0",
    platform_type = "macos",
)
```

## Configurations that do not work around it

| Attempt | Result |
|---|---|
| Reorder `--macos_cpus` | Moves the defect to the other slice |
| `apple_universal_binary` with `forced_cpus` | Same defect, first forced CPU wins |
| `--incompatible_enable_apple_toolchain_resolution=true` | `--macos_cpus` ignored, produces a thin arm64 binary |
| ... plus `--apple_platforms=@apple_support//platforms:macos_x86_64,@apple_support//platforms:macos_arm64` | Still thin arm64 |

Pinning rules_apple to 4.3.3 is the only workaround we found.

## Possible reason this went unnoticed

The plist tests in `test/starlark_tests/macos_command_line_application_tests.bzl` all
run with `"//command_line_option:macos_cpus": "x86_64"`, so the embedded-plist
assertions never exercise a split build. A universal variant of
`{}_merged_info_plist_binary_contents_test` asserting the section on each slice would
cover this.

## Suggested fix

Dropping the incoming rule transition from the two plist rules restores per-split
configuration and fixes every case above:

```diff
 macos_binary_infoplist = rule(
     implementation = _macos_binary_infoplist_impl,
-    cfg = transition_support.apple_rule_transition,
     attrs = dicts.add(
         apple_support.platform_constraint_attrs(),
         rule_attrs.common_tool_attrs(),
-        rule_attrs.custom_transition_allowlist_attr(),

 macos_command_line_launchdplist = rule(
     implementation = _macos_command_line_launchdplist_impl,
-    cfg = transition_support.apple_rule_transition,
     attrs = dicts.add(
         apple_support.platform_constraint_attrs(),
         rule_attrs.common_tool_attrs(),
-        rule_attrs.custom_transition_allowlist_attr(),
```

The allowlist attr has to go with the transition, otherwise Bazel reports "Unused
function-based split transition allowlist".

The part of #2846 that mattered in this file, `apple_platform_info_from_rule_ctx(ctx)`
replacing `platform_type_string`, still works without the transition: both rules are
only reached as deps inside the binary rule's split, so the incoming configuration is
already an Apple platform configuration. Building `//:tool.merged_infoplist` directly,
with or without `--macos_cpus`, also still works.

Verified against 4.5.3:

* both slices get the plist, in both `--macos_cpus` orderings
* single-architecture builds unchanged
* `launchdplists` and `apple_universal_binary` fixed as well
* `//test/starlark_tests:macos_command_line_application` passes 18 of 18
* the `macos_application` and `apple_universal_binary` suites give 34 passes and 7
  build failures both with and without the patch (a broken
  `ccinfo_dylibs_mylib-shared-with-rpath` genrule, unrelated to this change)

Two things you may want to weigh:

* This restores building the merged plist once per split. The two outputs are
  byte-identical, so it is redundant work rather than divergent output, but if you
  would rather keep the single action that 4.4.0 introduced, the alternative is to
  keep the collapse and propagate that one plist's linkopt into every split.
* More broadly, `_apple_rule_base_transition_impl` discarding everything after `[0]`
  may be wrong for any rule reachable inside a split, not just these two. Preserving
  an incoming Apple split configuration would address the whole class, but could
  change behavior for nested bundles whose `platform_type` differs from the parent.

## Environment

| | |
|---|---|
| Bazel | 8.6.0 |
| Xcode | 26.4 (17E192) |
| macOS | 26.5.2 (25F84) |
| Host arch | arm64 |
